File: javascript-http-sam/src/api/resources.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

// Implementation of the API backend for resources

const { DynamoDB } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocument } = require("@aws-sdk/lib-dynamodb");
const { metricScope } = require("aws-embedded-metrics");
const uuid = require("uuid");
ddbClient = new DynamoDB({});

// Check if code is running in SAM Local environment and disable X-Ray in such case
if (!process.env.AWS_SAM_LOCAL) {
  const AWSXRay = require("aws-xray-sdk-core");
  ddbClient = AWSXRay.captureAWSv3Client(new DynamoDB({}));
}

const dynamo = DynamoDBDocument.from(ddbClient);
const tableName = process.env.RESOURCES_TABLE;

exports.handler = metricScope(metrics => async (event, context) => {
  let body;
  let statusCode = 200;
  const headers = {
    "Content-Type": "application/json"
  };

  // Initialize putting common business metrics using EMF
  var metricPayload = {}
  metrics.putDimensions({ Service: "Resources" });
  metrics.putMetric("ProcessedResources", 1, "Count");
  metrics.setProperty("requestId", event.requestContext.requestId);
  metrics.setProperty("routeKey", event.routeKey);

  try {
    switch (event.routeKey) {
      // Get all resources
      case "GET /locations/{locationid}/resources":
        // add business metrics for the route
        metricPayload.operation = "GET";
        metricPayload.locationid = event.pathParameters.locationid;
        // get data from the database
        body = await dynamo.query({
          TableName: tableName,
          IndexName: "locationidGSI",
          KeyConditionExpression: "locationid = :locationid",
          ExpressionAttributeValues: {
            ':locationid': event.pathParameters.locationid
          }
        });
        break;
      // Resource CRUD operations
      case "GET /locations/{locationid}/resources/{resourceid}":
        // add business metrics for the route
        metricPayload.operation = "GET";
        metricPayload.resourceid = event.pathParameters.resourceid;
        metricPayload.locationid = event.pathParameters.locationid;
        // get data from the database
        body = await dynamo
          .get({
            TableName: tableName,
            Key: {
              resourceid: event.pathParameters.resourceid
            }
          });
        break;
      case "DELETE /locations/{locationid}/resources/{resourceid}":
        // add business metrics for the route
        metricPayload.operation = "DELETE";
        metricPayload.resourceid = event.pathParameters.resourceid;
        metricPayload.locationid = event.pathParameters.locationid;
        // delete item in the database
        await dynamo
          .delete({
            TableName: tableName,
            Key: {
              resourceid: event.pathParameters.resourceid
            }
          });
        body = {};
        break;
      case "PUT /locations/{locationid}/resources":
        let requestJSON = JSON.parse(event.body);
        // generate unique id if it isn't present in the request
        let id = requestJSON.resourceid || uuid.v1();
        let newitem = {
          resourceid: id,
          locationid: event.pathParameters.locationid,
          timestamp: new Date().toISOString(),
          description: requestJSON.description,
          name: requestJSON.name,
          type: requestJSON.type
        }
        // add business metrics for the route
        metricPayload.operation = "PUT";
        metricPayload.resourceid = id;
        metricPayload.locationid = event.pathParameters.locationid;
        // update the database
        await dynamo
          .put({
            TableName: tableName,
            Item: newitem
          });
        body = newitem;
        break;
      default:
        throw new Error(`Unsupported route: "${event.routeKey}"`);
    }
  } catch (err) {
    statusCode = 400;
    body = err.message;
    console.error(err.message);
  } finally {
    // Add route specific business metrics
    metrics.setProperty("Payload", metricPayload);
    body = JSON.stringify(body);
  }

  return {
    statusCode,
    body,
    headers
  };
});

File: python-http-cdk/src/api/resources.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

import boto3
import json
import os
import uuid
from datetime import datetime

from aws_embedded_metrics import metric_scope
from aws_xray_sdk.core import patch_all

patch_all()

RESOURCES_TABLE = os.getenv('RESOURCES_TABLE', None)
dynamodb = boto3.resource('dynamodb')
ddbTable = dynamodb.Table(RESOURCES_TABLE)


@metric_scope
def lambda_handler(event, context, metrics):
    route_key = event['routeKey']

    response_body = {'Message': 'Unsupported route'}
    status_code = 400
    headers = {'Content-Type': 'application/json'}

    # Put common business metrics using EMF
    metric_payload = {}
    metrics.put_dimensions({'Service': 'Resources'})
    metrics.put_metric('ProcessedResources', 1, 'Count')
    metrics.set_property('requestId', event['requestContext']['requestId'])
    metrics.set_property('routeKey', event['routeKey'])

    try:
        # Get all resources
        if route_key == 'GET /locations/{locationid}/resources':
            metric_payload['operation'] = 'GET'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            ddb_response = ddbTable.query(
                IndexName='locationidGSI',
                KeyConditionExpression='locationid = :locationid',
                ExpressionAttributeValues={
                    ':locationid': event['pathParameters']['locationid']
                }
            )
            response_body = ddb_response['Items']
            status_code = 200
        # Resource CRUD operations
        if route_key == 'GET /locations/{locationid}/resources/{resourceid}':
            metric_payload['operation'] = 'GET'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            metric_payload['resourceid'] = event['pathParameters']['resourceid']
            ddb_response = ddbTable.get_item(
                Key={'resourceid': event['pathParameters']['resourceid']}
            )
            if 'Item' in ddb_response:
                response_body = ddb_response['Item']
            else:
                response_body = {}
            status_code = 200
        if route_key == 'DELETE /locations/{locationid}/resources/{resourceid}':
            metric_payload['operation'] = 'DELETE'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            metric_payload['resourceid'] = event['pathParameters']['resourceid']
            ddbTable.delete_item(
                Key={'resourceid': event['pathParameters']['resourceid']}
            )
            response_body = {}
            status_code = 200
        if route_key == 'PUT /locations/{locationid}/resources':
            request_json = json.loads(event['body'])
            request_json['locationid'] = event['pathParameters']['locationid']
            request_json['timestamp'] = datetime.now().isoformat()
            if 'resourceid' not in request_json:
                request_json['resourceid'] = str(uuid.uuid1())
            metric_payload['operation'] = 'DELETE'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            metric_payload['resourceid'] = request_json['resourceid']
            ddbTable.put_item(
                Item=request_json
            )
            response_body = request_json
            status_code = 200
    except Exception as err:
        status_code = 400
        response_body = {'Error:': str(err)}
        print(str(err))
    metrics.set_property("Payload", metric_payload)
    return {
        'statusCode': status_code,
        'body': json.dumps(response_body),
        'headers': headers
    }
